### PR TITLE
Bug 1241136  -copy the shortened link in Telemetry dashboards

### PR DESCRIPTION
**Pull request by hmntptwl**
Added document.execCommand('copy') in the response handler so that shortened link in telemetry dashboards can be copied without a need to remember to hit ^C.

some background: 
- [https://jsfiddle.net/homz0zv0/2/](https://jsfiddle.net/homz0zv0/2/)

- [https://developer.mozilla.org/en-US/docs/Web/API/Document/execCommand](https://developer.mozilla.org/en-US/docs/Web/API/Document/execCommand)

- [Browser Support](http://caniuse.com/#feat=document-execcommand)

### CODE REVIEW

**Comment by chutten** [comment]
First appearances look good, but there are a couple things you can do to make it better. First and foremost, writing a good commit message is necessary. A good format we follow is
```
bug <###> - Made foo do bar

Foo previously didn't do bar because baz took care of it.
```

Essentially: bug number in the subject line, a quick summary of what the commit did ("updated dashboard.js" is something we can tell from git, you don't need to tell us that. Try something like "Copy the shortened link"), and then any background information or reasons why you did it one way instead of another (For instance, I'm not clear on why you didn't just call document.execCommand("copy"); on its own line instead of using the jQuery wrapper. Also, what's the browser support like for execCommand?)

And, if you want to make things _really_ easy for me, you can push this to your gh-pages branch and have github host it for you so I can review a live preview of the code :)

**Comment by hmntptwl** [comment]
Yea I will do this :) & thanks for the review.

File: src/dashboards.js [hunk 1]
```diff
@@ -164,6 +164,7 @@ $(document)
               .show()
               .val(shortUrl)
               .focus();
+               document.execCommand('copy');
```
Comment (chutten) [inline src/dashboards.js]: Bring the indentation of this line back to the correct level, please.

**Review by chutten — CHANGES_REQUESTED** [review]
Please squash these two commits together into one and address the other comment. 

The commit message is much better. I would recommend thinking about good commit message bodies as answering the question "Why?" instead of the question "What?". In my experience reasons tend to be more useful when looking through commit history, and "What?" is usually easy to figure out by looking at the patch contents and the commit message subject line.

Not something you should feel you must do here, but a good habit to develop.

File: src/dashboards.js (hunk 1 above)
Comment (hmntptwl) [reply src/dashboards.js]: Well I am learning best of open source practices. :)

**Comment by hmntptwl** [comment]
I have hosted source code here: https://github.com/hemant1612/hemant1612.github.io
However  [https://hemant1612.github.io/new-pipeline/dist.html](https://hemant1612.github.io/new-pipeline/dist.html) is not copying the generated url.
I think its due to a console error.
And `document.execCommand('copy')` needs to be updated in https://github.com/hemant1612/hemant1612.github.io/blob/master/new-pipeline/src/dashboards.js#L170.
I added `alert()`  just for debugging purposes.
Trying to simulate the process [here](https://jsfiddle.net/rgc6kj3g/1/) works fine.

**Comment by chutten** [comment]
Ah, the good ol' `document.execCommand(‘cut’/‘copy’) was denied because it was not called from inside a short running user-generated event handler.`

This is probably because the `execCommand` call is inside the xhr's handler, not the button's. If you force the XHR to be synchronous and pull the `execCommand` line out into the button's handler instead, does it work?

**Comment by hmntptwl** [comment]
I strongly agree with you here.However still it is not working.
https://github.com/hemant1612/hemant1612.github.io/blob/master/new-pipeline/src/dashboards.js#L170

**Comment by chutten** [comment]
Slapping a breakpoint on the `execCommand` is showing that it is being called before the selection is happening. So it is working! But there isn't anything yet selected for it to copy.

Make sure the XHR is completing synchronously, I guess? And if it is, make sure the input's value is being set at the right time and focus is being processed correctly.

**Comment by hmntptwl** [comment]
Finally gotta it work.

The way I tried to implement it earlier was not working because 

- Copy and cut commands triggered through a scripting API will only affect the contents of the real clipboard if the event is dispatched from an event that is trusted and triggered by the user, or if the implementation is configured to allow this. [W3 SPECS](https://www.w3.org/TR/clipboard-apis/#integration-with-rich-text-editing-apis)

Here a `$(window).blur() or $(document).click() event` can save the life.
Or to populate the input by the time user has clicked which I have done [here](https://github.com/hemant1612/hemant1612.github.io/blob/master/new-pipeline/src/dashboards.js#L130).

File: new-pipeline/src/dashboards.js [hunk 2]
```diff
@@ -127,51 +127,49 @@ $(document)
...
+        shortUrl = response.results[longUrl].shortUrl;
+        if (shortUrl.indexOf(":") === 4) {
+          shortUrl = "https" + shortUrl.substring(4);
+        }
+        input.val(shortUrl);
+        trigger.attr('disabled', false);
+      }
+    })
+  };
+getIt();
```
Comment (chutten) [inline new-pipeline/src/dashboards.js]: This means you're getting the short url on the pageload, not after the user has made their selection (which changes the url parameters).

File: new-pipeline/src/dashboards.js [hunk 3]
```diff
@@ -127,51 +127,49 @@ $(document)
...
-            $this.parents(".permalink-control")
-              .find("input")
-              .show()
-              .val(shortUrl)
-              .focus();
-            document.execCommand('copy');
-          }
-        });
-      });
+   $(".permalink-control")
```
Comment (chutten) [inline new-pipeline/src/dashboards.js]: The indentation here is wrong

File: src/dashboards.js [hunk 4]
```diff
@@ -164,6 +164,7 @@ $(document)
               .show()
               .val(shortUrl)
               .focus();
+            document.execCommand('copy');
```
Comment (chutten) [inline src/dashboards.js]: Don't bother with src/dashboards.js. It is deprecated.

File: new-pipeline/src/dashboards.js [hunk 5]
```diff
@@ -127,50 +127,49 @@ $(document)
...
-            }
-            $this.parents(".permalink-control")
-              .find("input")
-              .show()
-              .val(shortUrl)
-              .focus();
-          }
-        });
-      });
+   $(".permalink-control")
```
Comment (chutten) [inline new-pipeline/src/dashboards.js]: The indentation here is wrong.

**Review by chutten — CHANGES_REQUESTED** [review]
I see what the problem is. `dataType: "jsonp"`

There is no such thing as a synchronous jsonp request.

So either we get the short urls every time the url changes and call copy when the link button is clicked (which is not dissimilar to what you have here, just with more calls to `getIt()`)

Or we go back to the person asking for this feature and say "Sorry, can't do it"

Personally, I'm aiming for the second. Unless there _is_ a synchronous way to shorten a url using the bitly API.

**Comment by hmntptwl** [comment]
I think we can't make API responses synchronous at least in any way that won't freeze the browser nd network connections are fundamentally asynchronous . I have asked  bitly developers  about that .Hope they reply soon..  :)
What about the first option we have !

**Comment by chutten** [comment]
I think a normal XHR using the modern [v3 bit.ly API](http://dev.bitly.com/links.html#v3_shorten) can be made synchronous as it serves its JSON response in the response body. I'm okay if this is a source of lag, as it is in response to a user interaction (so we're allowed to be a little slower) and bit.ly's service tends to be pretty rapid.

File: src/dashboards.js (hunk 4 above)
Comment (hmntptwl) [inline src/dashboards.js]: Please review the changes :)
[https://hemant1612.github.io](https://hemant1612.github.io)

**Comment by chutten** [comment]
The system works!

Can you squash all these commits into a single one with an excellent commit message? Then we can get this in and live on telemetry.mozilla.org

**Comment by hmntptwl** [comment]
sure :).

**Comment by chutten** [comment]
Oh! And in all this I never noticed that you submitted this pull request against chutten/telemetry-dashboard instead of mozilla/telemetry-dashboard. Could you move it over there?

**Comment by hmntptwl** [comment]
LGTM

**Pull request closed without merging**